Consider merging 478ba2c (#8306) into release-3.2

Sometimes we got  etcd error: `rpc error: code = Unknown desc = auth: revision in header is old`

#8306  fixes this issue in master which does not appear in release-3.2.

Yes, i can always reproduce in my env. Below is steps.

1. starting a etcd server in terminal 1:
`$ etcd`

2. add `root` user and enable auth in terminal 2:
```bash
$ export ETCDCTL_API=3
$ echo root | etcdctl user add root --interactive=false
$ etcdctl auth enable
```

3. run [gist](https://gist.github.com/zhaoyao/1563a218ea0df666786435a7e821e3f6) in terminal 3

here is my output:
```bash
panic: rpc error: code = Unknown desc = auth: revision in header is old

goroutine 39 [running]:
main.test(0xc420098f20, 0xc42019e180, 0x1)
       ***/main.go:46 +0x815
created by main.main
        ***/main.go:67 +0x213
panic: rpc error: code = Unknown desc = auth: revision in header is old

goroutine 45 [running]:
main.test(0xc420098f20, 0xc42019e180, 0x7)
        ***/main.go:46 +0x815
created by main.main
        ***/main.go:67 +0x213
exit status 2
```

I have tested in both master and release-3.2 branch

master works fine
```
etcd Version: 3.2.0+git
Git SHA: fa32a85e6
Go Version: go1.8.3
Go OS/Arch: darwin/amd64
```

but release-3.2 will panic
```
etcd Version: 3.2.5+git
Git SHA: fa0f27878
Go Version: go1.8.3
Go OS/Arch: darwin/amd64

--- a/integration/v3_auth_test.go
+++ b/integration/v3_auth_test.go
@@ -15,6 +15,8 @@
 package integration
 
 import (
+	"fmt"
+	"sync"
 	"testing"
 	"time"
 
@@ -291,4 +293,42 @@ func TestV3AuthNonAuthorizedRPCs(t *testing.T) {
 	if !eqErrGRPC(err, rpctypes.ErrGRPCUserEmpty) {
 		t.Fatalf("could put key (%v), it should cause an error of permission denied", respput)
 	}
+}
+
+func TestV3AuthOldRevConcurrent(t *testing.T) {
+	defer testutil.AfterTest(t)
+	clus := NewClusterV3(t, &ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	authSetupRoot(t, toGRPC(clus.Client(0)).Auth)
+
+	c, cerr := clientv3.New(clientv3.Config{
+		Endpoints:   clus.Client(0).Endpoints(),
+		DialTimeout: 5 * time.Second,
+		Username:    "root",
+		Password:    "123",
+	})
+	testutil.AssertNil(t, cerr)
+	defer c.Close()
+
+	var wg sync.WaitGroup
+	f := func(i int) {
+		defer wg.Done()
+		role, user := fmt.Sprintf("test-role-%d", i), fmt.Sprintf("test-user-%d", i)
+		_, err := c.RoleAdd(context.TODO(), role)
+		testutil.AssertNil(t, err)
+		_, err = c.RoleGrantPermission(context.TODO(), role, "", clientv3.GetPrefixRangeEnd(""), clientv3.PermissionType(clientv3.PermReadWrite))
+		testutil.AssertNil(t, err)
+		_, err = c.UserAdd(context.TODO(), user, "123")
+		testutil.AssertNil(t, err)
+		_, err = c.Put(context.TODO(), "a", "b")
+		testutil.AssertNil(t, err)
+	}
+	// needs concurrency to trigger
+	numRoles := 2
+	wg.Add(numRoles)
+	for i := 0; i < numRoles; i++ {
+		go f(i)
+	}
+	wg.Wait()
 }

--- a/pkg/testutil/assert.go
+++ b/pkg/testutil/assert.go
@@ -54,5 +54,9 @@ func AssertFalse(t *testing.T, v bool, msg ...string) {
 }
 
 func isNil(v interface{}) bool {
-	return v == nil || reflect.ValueOf(v).IsNil()
+	if v == nil {
+		return true
+	}
+	rv := reflect.ValueOf(v)
+	return rv.Kind() != reflect.Struct && rv.IsNil()
 }

